eth/beacon/deposit_helpers.py

```diff
@@ -0,0 +1,160 @@
+        proof_of_possession=EMPTY_SIGNATURE,
+    )
+
+    is_valid_signature = bls.verify(
+        pubkey=pubkey,
+        # TODO: change to hash_tree_root(deposit_input) when we have SSZ tree hashing
+        message=deposit_input.root,
+        signature=proof_of_possession,
+        domain=get_domain(
+            state.fork_data,
+            state.slot,
+            SignatureDomain.DOMAIN_DEPOSIT,
+        ),
+    )
+
+    if not is_valid_signature:
+        raise ValidationError(
+            "BLS signature verification error"
+        )
+
+
+def add_pending_validator(state: BeaconState,
+                          validator: ValidatorRecord,
+                          zero_balance_validator_ttl: int) -> Tuple[BeaconState, int]:
+    """
+    Add a validator to the existing minimum empty validator index or
+    append to ``validator_registry``.
+    """
+    # Check if there's empty validator index in `validator_registry`
+    try:
+        index = get_min_empty_validator_index(
+            state.validator_registry,
+            state.slot,
+            zero_balance_validator_ttl,
+        )
+    except MinEmptyValidatorIndexNotFound:
+        index = None
+
+    if index is None:
+        # Append to the validator_registry
+        validator_registry = state.validator_registry + (validator,)
+        state = state.copy(
+            validator_registry=validator_registry,
+        )
+        index = len(state.validator_registry) - 1
+    else:
+        # Use the empty validator index
+        state = state.update_validator(index, validator)
+
+    return state, index
+
+
+def process_deposit(*,
+                    state: BeaconState,
+                    pubkey: int,
+                    deposit: int,
+                    proof_of_possession: bytes,
+                    withdrawal_credentials: Hash32,
+                    randao_commitment: Hash32,
+                    zero_balance_validator_ttl: int) -> Tuple[BeaconState, int]:
+    """
+    Process a deposit from Ethereum 1.0.
+    """
+    validate_proof_of_possession(
+        state,
+        pubkey,
+        proof_of_possession,
+        withdrawal_credentials,
+        randao_commitment,
+    )
+
+    validator_pubkeys = tuple([v.pubkey for v in state.validator_registry])
+    if pubkey not in validator_pubkeys:
+        validator = ValidatorRecord.get_pending_validator(
+            pubkey=pubkey,
+            withdrawal_credentials=withdrawal_credentials,
+            randao_commitment=randao_commitment,
+            balance=deposit,
+            latest_status_change_slot=state.slot,
+        )
+
+        state, index = add_pending_validator(
+            state,
+            validator,
+            zero_balance_validator_ttl,
+        )
+    else:
+        # Top-up - increase balance by deposit
+        index = validator_pubkeys.index(pubkey)
+        validator = state.validator_registry[index]
+
+        if validator.withdrawal_credentials != validator.withdrawal_credentials:
```

Comment:
This conditional looks broken.  Seems like you are comparing the same value with itself.  Is one of these supposed to be just `withdrawal_credentials`.

Maybe add a test that covers this code path since this code path probably isn't being exercised in the tests right now?

Reply:
Yeah, should be checking that the existing validator with the pubkey `pubkey` has the same `withdrawal_credentials` that were passed in with the deposit